### PR TITLE
PRO-4543 settings module

### PR DESCRIPTION
**Pull request by myovchev**
## Summary

Add `@apostrophecms/settings` module. Implement settings configuration per logged in user (UI and backend). 
Although fully functional, the final UI will be done in a following ticket. 

## What are the specific steps to test this change?

Use Testbed branch `feature-user-settings`, link Apostrophe locally while in the current branch. Boot the app, force rebuild (`CI=1 APOS_DEV=1 npx nodemon`)

"Manage Settings" should be visible above the "Logout" menu item (top right user menu). Clicking it should open "Manage Settings" modal. "UI Language" schema should be visible. Changing it should change the current user "UI Language" field (can be checked in the users piece). Re-opening the settings manager should show the new "UI language" setting by default.

## What kind of change does this PR introduce?
*(Check at least one)*

- [ ] Bug fix
- [x] New feature
- [ ] Refactor
- [ ] Documentation
- [ ] Build-related changes
- [ ] Other

## Make sure the PR fulfills these requirements:

- [x] It includes a) the existing issue ID being resolved, b) a convincing reason for adding this feature, or c) a clear description of the bug it resolves
- [ ] The changelog is updated
- [ ] Related documentation has been updated
- [x] Related tests have been updated

If adding a new feature without an already open issue, it's best to open a **feature request issue** first and wait for approval before working on it.

**Other information:**

### CODE REVIEW

File: modules/@apostrophecms/settings/index.js [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+          if (forbiddenFields.includes(name)) {
+            throw new Error(`[@apostrophecms/settings] The field "${name}" is forbidden.`);
+          }
+        }
+      },
+
+      // Get subform fields by subform name.
+      getSubformSchema(name) {
+        const subform = self.options.subforms[name];
+        if (!subform) {
```
Comment (boutell) [inline modules/@apostrophecms/settings/index.js:97]: How about a typeof check as well, making sure it is an object and has a fields array property?

File: modules/@apostrophecms/settings/index.js [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+            user: true
+          }
+        );
+      },
+
+      addSettingsModal() {
+        if (!self.hasSchema()) {
+          return;
+        }
+        self.apos.modal.add(
```
Comment (boutell) [inline modules/@apostrophecms/settings/index.js:127]: 👍 I had the signup module on the brain when I suggested doing it the hard way 😆

File: modules/@apostrophecms/settings/index.js [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+          throw self.apos.error('notfound');
+        }
+
+        const values = {
+          _id: user._id
+        };
+        for (const field of self.userSchema) {
+          values[field.name] = user[field.name];
+        }
+        return values;
```
Comment (boutell) [inline modules/@apostrophecms/settings/index.js:164]: Slightly unusual for REST to return an object at this level, but it makes sense here. 👍

File: modules/@apostrophecms/settings/index.js [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+      }
+    };
+  },
+
+  apiRoutes(self) {
+    return {
+      patch: {
+        ':subform': async (req) => {
```
Comment (boutell) [inline modules/@apostrophecms/settings/index.js:172]: Nice.

File: modules/@apostrophecms/settings/index.js (hunk 1 above)
Comment (myovchev) [reply modules/@apostrophecms/settings/index.js:97]: Good catch! I'll do that now.

File: modules/@apostrophecms/settings/index.js [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+        }
+        return schema;
+      },
+
+      // Validate that the fields configured in the settings module exist in the
+      // user schema and are not forbidden.
+      // XXX Temporary (Phase 2,3) extend the forbidden protected fields.
+      validateSettingsSchema(settingsFieldNames, userSchema) {
+        const forbiddenFields = [
+          'role',
```
[the remark this answers is not in the document]
Comment (myovchev) [reply modules/@apostrophecms/settings/index.js:78]: We can introduce a public options (and merge with it) once the need is identified.

File: modules/@apostrophecms/settings/index.js (hunk 1 above)
Comment (myovchev) [reply modules/@apostrophecms/settings/index.js:97]: No need to check here, instead we are doing it on a boot time (malformed `fields`). Additionally we throw on unknown field in `fields`. So it's a boot check and panic, getting it later is completely safe (should be).

**Review by boutell — APPROVED** [review]

**Pull request merged**